src/hotspot/share/opto/vmaskloop.cpp

```diff
@@ -0,0 +1,975 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "precompiled.hpp"
+#include "opto/addnode.hpp"
+#include "opto/convertnode.hpp"
+#include "opto/vmaskloop.hpp"
+
+//        L O O P   V E C T O R   M A S K   T R A N S F O R M A T I O N
+// ============================================================================
+
+// -------------------------------- Constructor -------------------------------
+VectorMaskedLoop::VectorMaskedLoop(PhaseIdealLoop* phase) :
+  _phase(phase),
+  _igvn(&(phase->_igvn)),
+  _arena(phase->C->comp_arena()),
+
+  _lpt(nullptr),
+  _cl(nullptr),
+  _cle(nullptr),
+  _iv(nullptr),
+
+  _core_set(_arena),
+  _body_set(_arena),
+  _body_nodes(_arena, 32, 0, nullptr),
+  _rpo_idx(_arena, 32, 0, 0),
+  _elem_bt(_arena, 32, 0, T_ILLEGAL),
+  _stmts(_arena, 2, 0, nullptr),
+  _swptrs(_arena, 8, 0, nullptr),
+  _size_stats(_arena)
+{}
+
+// ------------------- Entry function of vector masked loop -------------------
+void VectorMaskedLoop::try_vectorize_loop(IdealLoopTree* lpt) {
+  assert(UseMaskedLoop, "Option should be enabled");
+  assert(lpt->is_counted(), "Loop must be counted");
+  assert(lpt->is_innermost(), "Loop must be innermost");
+
+  CountedLoopNode *cl = lpt->_head->as_CountedLoop();
+  // Skip if loop is already vector masked
+  if (cl->is_vector_masked()) return;
+  // Skip non-post loop
+  if (!cl->is_post_loop()) return;
+  // Skip malformed counted loop
+  if (!cl->is_valid_counted_loop(T_INT)) return;
+  // Skip loop if stride is unsupported
+  if (abs(cl->stride_con()) != 1) return;
+  // Skip loop with control flow
+  if (cl->loopexit()->in(0) != cl) return;
+  // Skip if some loop operations are pinned to the backedge
+  if (cl->back_control()->outcnt() != 1) return;
+
+  // Init data structures and collect loop nodes
+  init(lpt);
+  if (!collect_loop_nodes()) return;
+
+  // Collect loop statements and analyze vectorizability
+  if (!collect_statements()) return;
+  if (!analyze_vectorizability()) return;
+
+  // Try creating a vector mask with the smallest vector element size
+  const TypeVectMask* t_vmask = create_vector_mask_type();
+  if (t_vmask == nullptr || !t_vmask->isa_vectmask()) return;
+
+  // Transform the loop and set flags
+  transform_loop(t_vmask);
+  cl->mark_loop_vectorized();
+  cl->mark_vector_masked();
+  _phase->C->set_max_vector_size(MaxVectorSize);
+  trace_msg(nullptr, "Loop is vector masked");
+}
+
+// ----------------------------------- Init -----------------------------------
+void VectorMaskedLoop::init(IdealLoopTree* lpt) {
+  // Set current loop info
+  _lpt = lpt;
+  _cl = lpt->_head->as_CountedLoop();
+  _cle = _cl->loopexit();
+  _iv = _cle->phi();
+
+  // Reset data structures
+  _core_set.clear();
+  _body_set.clear();
+  _body_nodes.clear();
```

Comment:
Would it make sense to somehow reserve the space, so that we do not allocate multiple times when growing these data structures later?

Reply:
Could you elaborate how to do such reservation in C2? Just allocation with some larger sizes at the beginning? Or any other examples to refer?

Reply:
I think what I have seen people do is just to `map` a high enough index value with `nullptr`. A bit hacky, but it ensures that the arrays underneath get grown sufficiently immediately. It would be nice to have some kind of `reserve` methods... but I don't think we have that.